agent/consul/gateways/controller_gateways.go

```diff
@@ -2,27 +2,208 @@ package gateways
+}
+
+func (r apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
+	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO not implemented
+	//httpRoutes, err := r.store.GetConfigEntriesByKind(structs.HTTPRoute)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	routes := []structs.BoundRoute{}
+	for _, r := range tcpRoutes {
+		if r == nil {
+			continue
+		}
+		routes = append(routes, r.(*structs.TCPRouteConfigEntry))
+	}
+	//TODO not implemented
+	//for _, r := range httpRoutes {
+	//	routes = append(routes, r.(*structs.HTTPRouteConfigEntry))
+	//}
+	return routes, nil
 }
 
 func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	entry, err := r.store.GetConfigEntry(req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+
+	if entry == nil {
+		r.logger.Info("cleaning up deleted gateway object", "request", req)
+		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           req.Name,
+			EnterpriseMeta: *req.Meta,
+		}); err != nil {
+			r.logger.Error("error cleaning up deleted gateway object", err)
+			return err
+		}
+		return nil
+	}
+
+	gatewayEntry := entry.(*structs.APIGatewayConfigEntry)
+	err = gatewayEntry.Validate()
+	if err != nil {
+		r.logger.Debug("persisting gateway status", "gateway", gatewayEntry)
+		if updateErr := r.store.UpdateStatus(gatewayEntry); err != nil {
+			return fmt.Errorf("%v: %v", err, updateErr)
+		}
+		return err
+	}
+
+	var boundGatewayEntry *structs.BoundAPIGatewayConfigEntry
+	boundEntry, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	if boundEntry == nil {
+		boundGatewayEntry = &structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           gatewayEntry.Name,
+			EnterpriseMeta: gatewayEntry.EnterpriseMeta,
+		}
+	} else {
+		boundGatewayEntry = boundEntry.(*structs.BoundAPIGatewayConfigEntry)
+	}
+
+	r.logger.Debug("started reconciling gateway")
+
+	routes, err := r.retrieveAllRoutesFromStore()
+
+	if err != nil {
+		return err
+	}
+
+	metaGateway := &gatewayMeta{
+		BoundGateway: boundGatewayEntry,
+		Gateway:      gatewayEntry,
+	}
+
+	//TODO is this needed? It seems to be for my tests
+	r.reconcileListeners(metaGateway)
+	fmt.Println(metaGateway.Gateway.Listeners)
+	fmt.Println(metaGateway.BoundGateway.Listeners)
+
+	boundGateways, routeErrors := BindRoutesToGateways(wrapSlice(metaGateway), routes...)
+
+	if len(boundGateways) > 1 {
+		err := fmt.Errorf("bind returned more gateways (%d) than it was given (1)", len(boundGateways))
+		r.logger.Error("API Gateway Reconciler failed to reconcile: %v", err)
+
+		return err
+	} else if len(routeErrors) > 0 && len(boundGateways) > 0 {
+		err := fmt.Errorf("bind returned no gateways and (%d) errors: %v", len(routeErrors), routeErrors)
```

Comment:
So this is currently saying, "if I have any route binding errors and successfully bound gateways", it shouldn't error out in this case. Really, the only thing that you'll want to do is a bounds check on `boundGateways` prior to indexing it at line 127. As the code is currently, if `routeErrors` > 0 and `boundGateways` == 0, then your code will panic.